level3.cc

```diff
@@ -0,0 +1,180 @@
+#include "level3.h"
+#include "pawn.h"
+#include "rook.h"
+#include "queen.h"
+#include "king.h"
+#include "knight.h"
+#include "bishop.h"
+
+Level3::Level3(ChessGame *subject, std::string name, std::vector<Piece*> *pieceArray, std::vector<Piece*> *oppArray):
+                Computer{subject, name, pieceArray, oppArray} {}
+
+
+//create second map (filteredMap)
+//loop through whole of pieceArray with iterator p, loop through p->LegalMoves() with iterator i
+//if i.second == 1 then add p:Box to second map and then randomly select from map
+
+bool Level3::pickMove() {
+    int size = 0;
+    std::cout << "Level 3 Computer makes its move." << std::endl;
+    std::map<Piece *, Box> filteredMap; // moves that are purely legal (doesn't put my king in check) and avoid capture
+    std::map<Piece *, Box> filteredMap2; // moves that are purely legal (doesn't put my king in check) and capture or checks
+    std::map<Piece *, Box> regularMap; // moves that are purely legal (doesn't put my king in check) and are not special (just a move)
+    std::vector<Box> toAvoid;
+    std::vector<Piece*> pieces = *pieceArray;
+    std::vector<Piece*> opponentPieces = *oppArray;
+    int arraySize = pieces.size();
+    int opponentArraySize = opponentPieces.size();
+
+    // looping through all of opponent's pieces
+    for (int i = 0; i < opponentArraySize; ++i) {
+        std::map<Box, int> opponentsLegalMoves = (*(*(opponentPieces[i])).getLegalMoves());
+        
+        // loop through the piece's legal moves
+        for (auto &move: opponentsLegalMoves) { 
+            // if opponent has a capture move add it to toAvoid (places to avoid) 
+            if (move.second == 1) {
+                toAvoid.push_back(move.first);
+            }
+        }
+    }
+
+    // checking is a move puts my king in check
+    for (int i = 0; i < arraySize; ++i) {
+        std::map<Box, int> potentialMoves = (*((*(pieces[i])).getLegalMoves()));
+        for (auto &move: potentialMoves) { 
```

Comment:
Quick question - is move the <Box, int> pair itself or an iterator (and thus a pointer?) to the pair in question?

Reply:
move is an iterator so we are dereferencing it so then it will be just a <Box, int> pair